### ISSUE TITLE
Some notes on the latest iteration.

### ISSUE THREAD

**Issue by Roger-Heathcote** [foundersandcoders/ws-session-management#42]
1. A lot of people were a bit confused about what foo and bar meant in the jwt docs and several went down rabbit holes.
2. A number of people mentioned they really would have liked a basic example of jwt.sign and jwt.verify as part of the tutorial.
3. Several teams weren't 100% clear that the example code was just for illustrative purposes and that the only part they had to provide to jwt.sign was the payload and key string.
4. It's also not obvious that verify throws deliberately when it receives an invalid token and that that error would need to be caught. 

While there is value in figuring out all this stuff by yourself it can take a long time. Only one team finished the first half this time, several of them got stuck and had to check the solutions, and nobody made it as far as the optional question. Comments welcome on whether we should make the workshop easier by addressing some/all of the above.

**Comment by oliverjam** [comment]
I agree the way this workshop is framed confuses people. There's a whole bunch of "this is sort of how signed tokens hypothetically would be implemented", then you just go and use the JWT library to do it and can actually ignore all that code.

I think this is because originally the challenges were reversed—you implemented your own HMAC function first, _then_ got to do it the "easy" way with the JWT library.

My suggestion would either be to drop the second challenge entirely (since it's unlikely they'll be implementing their own token signing at this point), or at least move all the HMAC example code to after the first challenge.

I also agree that the JWT docs suck, so we should have simple examples in the README (although that may make challenge 1 very easy, since pretty much the entire "challenge" is figuring out how to use sign/verify).

**Comment by Roger-Heathcote** [comment]
I agree examples might make it too easy, and there is value in practising how to extract the information you need from less than perfect docs. Maybe the solution is to add more clarifying text to the challenge to make it very explicit that...

1. The end goal is to sign a payload that, at minimum, contains a users unique ID.
2. The JWT library does that by implementing all of the above code for you
3. The challenge is to figure out how to use it to sign and verify by reading the docs.
4. The terms "foo" and "bar" are long standing, traditional placeholders in CS examples.

**Comment by oliverjam** [comment]
I might be rewriting this from scratch for FAC21, but those are great things to bear in mind if/when I do
